include/equation_system.hpp

```diff
@@ -0,0 +1,100 @@
+#ifndef ADJACENT_EQUATION_SYSTEM_HPP
+#define ADJACENT_EQUATION_SYSTEM_HPP
+
+#include <vector>
+#include <unordered_map>
+
+#include <xtensor/xtensor.hpp>
+#include "expression.hpp"
+#include "expression_vector.hpp"
+#include "gaussian_method.hpp"
+
+enum SolveResult {
+	OKAY,
+	DIDNT_CONVERGE,
+	REDUNDANT,
+	POSTPONE
+};
+
+using expr_ptr = std::shared_ptr<Expr>;
+
+class EquationSystem  {
+
+public:
+	bool is_dirty = true;
+	int max_steps = 20;
+	int drag_steps = 3;
+	bool revert_when_not_converged = true;
+
+	std::string stats;
+	bool dof_changed;
+
+
+	xt::xtensor<std::shared_ptr<Expr>, 2> J;
+	xt::xtensor<double, 2> A;
+	xt::xtensor<double, 2> AAT;
+	xt::xtensor<double, 1> B;
+	xt::xtensor<double, 1> X;
+	xt::xtensor<double, 1> Z;
+	xt::xtensor<double, 1> old_param_value;
+
+	std::vector<std::shared_ptr<Expr>> source_equations;
+	std::vector<Param<double>> parameters;
+
+	std::vector<std::shared_ptr<Expr>> equations;
+	std::vector<Param<double>> current_params;
+
+	std::unordered_map<Param<double>, Param<double>> subs;
+
+	void add_equation(const std::shared_ptr<Expr>& eq);
```

Comment:
I prefer to use camel naming like "addEquation". But also we can prefer codestyle used in FreeCAD since we are going to integrate. But doesn't matter what to use.